Detectors/TRD/simulation/src/Trap2CRU.cxx

```diff
@@ -48,86 +49,203 @@ namespace o2
+      std::stringstream ss;
+      ss << std::setw(2) << std::setfill('0') << sm;
+      std::string supermodule = ss.str();
+      outFileLink = o2::utils::concat_string(mOutputDir, "/", outPrefix, "_sm_", supermodule, outSuffix);
+    } else if (mFilePer == "cru") {
+      outFileLink = o2::utils::concat_string(mOutputDir, "/", outPrefix, "_cru_", std::to_string(mCruID), outSuffix);
+    } else if (mFilePer == "halfcru") {
+      outFileLink = o2::utils::concat_string(mOutputDir, "/", outPrefix, "_cru_", std::to_string(mCruID), trdside, outSuffix);
+    } else {
+      throw std::runtime_error("invalid option provided for file grouping");
+    }
+
+    //std::string outputFilelink = o2::utils::concat_string(prefix, "trd_cru_", std::to_string(mCruID), "_", trdside, "_", whichrun, ".raw");
+    LOG(info) << "registering links";
+
+    mWriter.registerLink(mFeeID, mCruID, mLinkID, mEndPointID, outFileLink);
+  }
+
+  openInputFiles();
+  sortDataToLinks();
+
+  if (mDigitsTree->GetEntries() != 1) {
+    LOG(fatal) << "more than one entry in digits tree " << mDigitsTree->GetEntries();
   }
-  mTrapRawFile = TFile::Open(inputFilename.data());
-  assert(mTrapRawFile != nullptr);
-  LOG(debug) << "Trap Raw file open " << inputFilename;
-  mTrapRawTree = (TTree*)mTrapRawFile->Get("o2sim");
-
-  mTrapRawTree->SetBranchAddress("TrapLinkRecord", &mLinkRecordsPtr);      // branch with the link records
-  mTrapRawTree->SetBranchAddress("RawTriggerRecord", &mTriggerRecordsPtr); // branch with the trigger records
-  mTrapRawTree->SetBranchAddress("TrapRaw", &mTrapRawDataPtr);             // branch with the actual incoming data.
-
-  for (int entry = 0; entry < mTrapRawTree->GetEntries(); entry++) {
-    mTrapRawTree->GetEntry(entry);
-    LOG(debug) << "Before Trigger Reord loop Event starts at:" << mTriggerRecords[0].getFirstEntry() << " and has " << mTriggerRecords[0].getNumberOfObjects() << " entries";
+  if (mTrackletsTree->GetEntries() != 1) {
+    LOG(fatal) << "more than one entry in tracklets tree " << mTrackletsTree->GetEntries();
+  }
+
+  for (int entry = 0; entry < mDigitsTree->GetEntries(); entry++) {
+    mDigitsTree->GetEntry(entry);
     uint32_t linkcount = 0;
-    for (auto trigger : mTriggerRecords) {
+  }
+  for (int entry = 0; entry < mTrackletsTree->GetEntries(); entry++) {
+    mTrackletsTree->GetEntry(entry);
+  }
+  // everything is passed as the first entry in the tree.
+  // and then work with the resulting vectors from the branches.
+  uint32_t linkcount = 0;
+  for (auto tracklettrigger : mTrackletTriggerRecords) {
+    //get the event limits from TriggerRecord;
+    uint32_t eventstart = tracklettrigger.getFirstEntry();
+    uint32_t eventend = tracklettrigger.getFirstEntry() + tracklettrigger.getNumberOfObjects();
+    LOG(debug) << "looping over tracklet trigger : " << eventstart << " -> " << eventend << " with bc of " << tracklettrigger.getBCData().bc;
+    for (auto digitstrigger : mDigitTriggerRecords) {
```

Comment:
there can be no digits if you choose not to use the digits in the conversion, as was the default before.
Even if the tracklets are empty the triggerrecord will be there but empty, it will still hit the digits. 
Yes I ummed and ahhhed which way to loop them.

For a system test, there is probably little reason to include the digits, they would get dumped before tracking in anycase (I think).